diff --git a/include/utils/SkBase64.h b/include/utils/SkBase64.h
index e01028543a..beddbd2c95 100644
--- a/include/utils/SkBase64.h
+++ b/include/utils/SkBase64.h
@@ -12,7 +12,7 @@
 
 #include <cstddef>
 
-struct SkBase64 {
+struct SK_API SkBase64 {
 public:
     enum Error {
         kNoError,
diff --git a/src/core/SkBitmapDevice.cpp b/src/core/SkBitmapDevice.cpp
index b497d690f7..5e4ee2b1ce 100644
--- a/src/core/SkBitmapDevice.cpp
+++ b/src/core/SkBitmapDevice.cpp
@@ -522,8 +522,8 @@ void SkBitmapDevice::onDrawGlyphRunList(SkCanvas* canvas,
                                         const sktext::GlyphRunList& glyphRunList,
                                         const SkPaint& initialPaint,
                                         const SkPaint& drawingPaint) {
-    SkASSERT(!glyphRunList.hasRSXForm());
-    LOOP_TILER( drawGlyphRunList(canvas, &fGlyphPainter, glyphRunList, drawingPaint), nullptr )
+    // SkASSERT(!glyphRunList.hasRSXForm());
+    // LOOP_TILER( drawGlyphRunList(canvas, &fGlyphPainter, glyphRunList, drawingPaint), nullptr )
 }
 
 void SkBitmapDevice::drawVertices(const SkVertices* vertices,
diff --git a/src/core/SkClipStackDevice.h b/src/core/SkClipStackDevice.h
index eff1f1a440..a8d6b4fe07 100644
--- a/src/core/SkClipStackDevice.h
+++ b/src/core/SkClipStackDevice.h
@@ -11,7 +11,7 @@
 #include "src/core/SkClipStack.h"
 #include "src/core/SkDevice.h"
 
-class SkClipStackDevice : public SkBaseDevice {
+class SK_API SkClipStackDevice : public SkBaseDevice {
 public:
     SkClipStackDevice(const SkImageInfo& info, const SkSurfaceProps& props)
         : SkBaseDevice(info, props)
diff --git a/src/core/SkDevice.h b/src/core/SkDevice.h
index e0fed94b9b..c7194f9c1c 100644
--- a/src/core/SkDevice.h
+++ b/src/core/SkDevice.h
@@ -54,7 +54,7 @@ struct SkStrikeDeviceInfo {
     const sktext::gpu::SDFTControl* const fSDFTControl;
 };
 
-class SkBaseDevice : public SkRefCnt, public SkMatrixProvider {
+class SK_API SkBaseDevice : public SkRefCnt, public SkMatrixProvider {
 public:
     SkBaseDevice(const SkImageInfo&, const SkSurfaceProps&);
 
diff --git a/src/core/SkFontPriv.h b/src/core/SkFontPriv.h
index 95ca905bf1..a31aba8e2b 100644
--- a/src/core/SkFontPriv.h
+++ b/src/core/SkFontPriv.h
@@ -16,7 +16,7 @@
 class SkReadBuffer;
 class SkWriteBuffer;
 
-class SkFontPriv {
+class SK_API SkFontPriv {
 public:
     /*  This is the size we use when we ask for a glyph's path. We then
      *  post-transform it as we draw to match the request.
